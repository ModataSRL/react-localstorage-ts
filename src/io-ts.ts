export * from "localvalue-ts/io-ts"
