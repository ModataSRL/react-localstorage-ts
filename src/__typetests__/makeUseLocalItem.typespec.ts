import * as t from "io-ts"
import {
  makeHooksFromStorage,
  makeStorageHooks,
  makeUseLocalItem,
} from "../useLocalStorage"
import { DateFromISOString } from "io-ts-types"
import { fromIoTsCodec } from "../io-ts"
import { createLocalStorage } from "localvalue-ts"

const ShapeCodec = t.type({ s: t.string, d: DateFromISOString })
type ShapeCodec = t.TypeOf<typeof ShapeCodec>

const CorrectCodec = fromIoTsCodec(ShapeCodec)

CorrectCodec.encode
const useShape = makeUseLocalItem("shape", CorrectCodec)

// @dts-jest:pass:snap It works with string encoding
makeUseLocalItem("sape", CorrectCodec)

// @dts-jest:fail:snap It doesn't work with non-string encoding
makeUseLocalItem("shape", ShapeCodec)

// @dts-jest:pass:snap You can pass a valid set of options
makeUseLocalItem("sape", CorrectCodec, {
  defaultValue: { s: "foo", d: new Date() },
  useMemorySore: false,
})

makeUseLocalItem("sape", CorrectCodec, {
  // @dts-jest:fail:snap You cannot pass an invalid set of options
  defaultValue: { foo: 123 },
  useMemorySore: false,
})

// @dts-jest:pass:snap resulting useShape typings are correct
useShape

// @dts-jest:fail:snap You cannot pass parameters to hooks
useShape("")

export const UnionCodec = t.union([t.literal("foo"), t.literal("baz")])
const CorrectUnionCodec = fromIoTsCodec(UnionCodec)

const hooks = makeStorageHooks({
  // @dts-jest:pass:snap It works with string encoding
  foo: CorrectCodec,
  union: CorrectUnionCodec,
})

const defaultShape = { s: "foo", d: new Date() }

makeStorageHooks({
  // @dts-jest:fail:snap It doesn't work with non-string encoding
  foo: ShapeCodec,
})

const hooksWithOptions = makeStorageHooks(
  {
    foo: CorrectCodec,
    union: CorrectUnionCodec,
  },
  // @dts-jest:pass:snap You can pass a valid set of options to store
  { useMemorySore: true, defaultValues: { foo: defaultShape } },
)

makeStorageHooks(
  {
    foo: CorrectCodec,
  },
  // @dts-jest:fail:snap You cannot pass an invalid set of options to store
  { useMemorySore: true, defaultValues: { fo: defaultShape } },
)

// @dts-jest:pass:snap store returns the correct type encoding
hooks

// @dts-jest:pass:snap store returns the correct type encoding
hooks.useUnion()

// @dts-jest:pass:snap storeWithOptions returns the correct type encoding
hooksWithOptions

// @dts-jest:pass:snap store returns the correct type encoding
hooksWithOptions.useUnion()

const store = createLocalStorage({
  foo: CorrectCodec,
  union: CorrectUnionCodec,
})

const storeWithOptions = createLocalStorage(
  {
    foo: CorrectCodec,
    union: CorrectUnionCodec,
  },
  { useMemorySore: true, defaultValues: { foo: defaultShape } },
)

const hooksFromStorage = makeHooksFromStorage(store)

const hooksFromStorageWithOptions = makeHooksFromStorage(storeWithOptions)

// @dts-jest:pass:snap store returns the correct type encoding
hooksFromStorage

// @dts-jest:pass:snap store returns the correct type encoding
hooksFromStorage.useUnion()

// @dts-jest:pass:snap storeWithOptions returns the correct type encoding
hooksFromStorageWithOptions

// @dts-jest:pass:snap store returns the correct type encoding
hooksFromStorageWithOptions.useUnion()
