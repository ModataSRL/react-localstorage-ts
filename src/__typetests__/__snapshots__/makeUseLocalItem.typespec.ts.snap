// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It doesn't work with non-string encoding (type) should match snapshot 1`] = `
"Argument of type 'TypeC<{ s: StringC; d: DateFromISOStringC; }>' is not assignable to parameter of type 'ValidCodec'.
  The types returned by 'encode(...)' are incompatible between these types.
    Type '{ s: string; d: string; }' is not assignable to type 'string'."
`;

exports[`It doesn't work with non-string encoding (type) should match snapshot 2`] = `"Type 'TypeC<{ s: StringC; d: DateFromISOStringC; }>' is not assignable to type 'Codec<any, string, any>'."`;

exports[`It works with string encoding (type) should match snapshot 1`] = `"LocalValueHook<Codec<Errors, string, { s: string; d: Date; }>>"`;

exports[`It works with string encoding (type) should match snapshot 2`] = `"Codec<Errors, string, { s: string; d: Date; }>"`;

exports[`You can pass a valid set of options (type) should match snapshot 1`] = `"LocalValueHook<Codec<Errors, string, { s: string; d: Date; }>>"`;

exports[`You can pass a valid set of options to store (type) should match snapshot 1`] = `"any"`;

exports[`You cannot pass an invalid set of options (type) should match snapshot 1`] = `
"Type '{ foo: number; }' is not assignable to type '{ s: string; d: Date; }'.
  Object literal may only specify known properties, and 'foo' does not exist in type '{ s: string; d: Date; }'."
`;

exports[`You cannot pass an invalid set of options to store (type) should match snapshot 1`] = `
"Type '{ fo: { s: string; d: Date; }; }' is not assignable to type 'Partial<{ foo: { s: string; d: Date; }; }>'.
  Object literal may only specify known properties, and 'fo' does not exist in type 'Partial<{ foo: { s: string; d: Date; }; }>'."
`;

exports[`You cannot pass parameters to hooks (type) should match snapshot 1`] = `"Expected 0 arguments, but got 1."`;

exports[`resulting useShape typings are correct (type) should match snapshot 1`] = `"LocalValueHook<Codec<Errors, string, { s: string; d: Date; }>>"`;

exports[`store returns the correct type encoding (type) should match snapshot 1`] = `"{ useFoo: LocalValueHook<Codec<Errors, string, { s: string; d: Date; }>>; useUnion: LocalValueHook<Codec<Errors, string, \\"foo\\" | \\"baz\\">>; }"`;

exports[`store returns the correct type encoding (type) should match snapshot 2`] = `"[item: LocalValue<Errors, \\"foo\\" | \\"baz\\">, setItem: (i: ValidLocalValue<\\"foo\\" | \\"baz\\">) => void]"`;

exports[`store returns the correct type encoding (type) should match snapshot 3`] = `"[item: LocalValue<Errors, \\"foo\\" | \\"baz\\">, setItem: (i: ValidLocalValue<\\"foo\\" | \\"baz\\">) => void]"`;

exports[`store returns the correct type encoding (type) should match snapshot 4`] = `"{ useFoo: LocalValueHook<Codec<Errors, string, { s: string; d: Date; }>>; useUnion: LocalValueHook<Codec<Errors, string, \\"foo\\" | \\"baz\\">>; }"`;

exports[`store returns the correct type encoding (type) should match snapshot 5`] = `"[item: LocalValue<Errors, \\"foo\\" | \\"baz\\">, setItem: (i: ValidLocalValue<\\"foo\\" | \\"baz\\">) => void]"`;

exports[`store returns the correct type encoding (type) should match snapshot 6`] = `"[item: LocalValue<Errors, \\"foo\\" | \\"baz\\">, setItem: (i: ValidLocalValue<\\"foo\\" | \\"baz\\">) => void]"`;

exports[`storeWithOptions returns the correct type encoding (type) should match snapshot 1`] = `"{ useFoo: LocalValueHook<Codec<Errors, string, { s: string; d: Date; }>>; useUnion: LocalValueHook<Codec<Errors, string, \\"foo\\" | \\"baz\\">>; }"`;

exports[`storeWithOptions returns the correct type encoding (type) should match snapshot 2`] = `"{ useFoo: LocalValueHook<Codec<Errors, string, { s: string; d: Date; }>>; useUnion: LocalValueHook<Codec<Errors, string, \\"foo\\" | \\"baz\\">>; }"`;
